draft-west-cookie-incrementalism.md

```diff
@@ -401,85 +404,281 @@ from "have the same registrable domain" (or "is an exact match for") to say
-{{RFC6265bis}} points out, this gives network attackers the ability to influence otherwise secured
-traffic by modifying user state that flows to secure origins, and, of course, insight into user
-behavior as securely-set cookies that lack the `Secure` attribute likewise flow from secure origins
-to non-secure variants.
+Finally, we don't ever want to allow a cookie to pass between schemes,
+given the huge security differences between them. So there should be no
+way for a server to specify that a given cookie should be sent to a
+different scheme.
 
-We should remedy this defect by storing a `scheme` component along with the cookie, and using that
-component in cookies' matching algorithms to ensure that secure and non-secure origins' state is
-clearly distinguishable and separate. This is accomplished as follows:
+How we accomplish this is as follows:
 
-First, alter the Storage Model defined in Section 5.4 of {{RFC6265bis}} by adding `scheme` to the
-list of fields the user agent stores about each cookie, and setting it when creating a cookie by
-altering step 2 of the same algorithm from:
+First, add the concept of port matching which helps to simplify checking
+if a cookie would match a port value. We can do that by adding a new
+section under 5.1 (this new section depends on the modification to 5.4
+below)
 
 ~~~
-2.   Create a new cookie with name cookie-name, value cookie-value.
-     Set the creation-time and the last-access-time to the current
-     date and time.
+5.1.5 Port matching 
+
+An integer port-matches a given cookie if any of the following
+conditions are true:
+  1. The cookie's host-only-flag is false.
+
+  2. The integer exactly matches the cookie's port value.
 ~~~
 
-to:
+Next, alter the storage model in Section 5.4 of {{RFC6265bis}} by
+adding "scheme" and "port" to the list of fields the user agent stores
+about each cookie, and setting them.
 
 ~~~
-2.   Create a new cookie with name cookie-name, value cookie-value.
-     Set the creation-time and the last-access-time to the current
-     date and time. Set the scheme to request-uri's origin's scheme
-     component.
+3.  Create a new cookie with name cookie-name, value cookie-value. Set
+    the creation-time and the last-access-time to the current date and
+    time.
 ~~~
 
-Likewise alter step 17 of the same algorithm from:
+to: 
 
 ~~~
-17.  If the cookie store contains a cookie with the same name,
-     domain, host-only-flag, and path as the newly-created cookie:
+3.  Create a new cookie with name cookie-name, value cookie-value. Set
+    the creation-time and the last-access-time to the current date and
+    time, set the scheme to the request-uri's origin's scheme
+    component, and set the port to the request-uri's origin's port
+    component.
 ~~~
 
-to:
+Because all cookies are now scoped by scheme it is unnecessary to
+protect a secure scheme's cookies from an insecure scheme's cookies and
+thus we can remove step 13.
```

Comment:
This would change behavior for sites mistakenly setting `secure` cookies on HTTP origins. Unless there's a good reason for changing that, it seems prudent to leave (admitedly ugly!) things like this alone.

Reply:
No good reason beyond it felt good to get rid of it. But you're right, we don't want to allow http to set `secure`.